rclcpp/include/rclcpp/executors/event_waitable.hpp

```
@@ -53,6 +53,13 @@ class EventWaitable : public rclcpp::Waitable
     return false;
   }
 
+  // Stub API: not used by EventsExecutor
+  std::shared_ptr<void>
+  take_data() final
+  {
+    return nullptr;
```

can we throw an error here?

---

Just waitables that are supposed to handle data (like subscription IPC) return something valid. The other waitable (like the event waitable) shoudl return nullptr and is not an error. The error should be thrown on the `execute` api, if `data` is not valid.

---

I should remove the comment `// Stub API: not used by EventsExecutor`, because now is used

---

there are 2 options: either the function throws an error or we don't implement it here.
If it can be used, the waitables should be allowed to modify it.

---

I tried to implement it only on `events_executor_notify_waitable.hpp` and removing a reference to `take_data` on `event_waitable.hpp` but the compiler complains, so I defined it here throwing an error

---

I think that you don't even need the declaration here at this point.